### PR TITLE
easyblock for Tau

### PR DESCRIPTION
**Pull request by boegel**
easyblock to simplify the complex easyconfig in https://github.com/hpcugent/easybuild-easyconfigs/pull/509, and deal with querying the toolchain

may need a bit of freshening up before it's good to go

cc @geimer, @berndmohr

### CODE REVIEW

File: easybuild/easyblocks/t/tau.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        # number of iterations: # backends + 1 (for basic)
+        iter_cnt = len(self.cfg['extra_backends']) + 1
+
+        self.cfg['configopts'] = [mpi_tmpl, openmp_tmpl, hybrid_tmpl] * iter_cnt
+
+        # custom prefix option for configure command
+        self.cfg['prefix_opt'] = '-prefix='
+
+        # installation command is 'make install clean'
+        self.cfg['installopts'] = 'clean'
```
Comment (wpoely86) [inline easybuild/easyblocks/t/tau.py]: wtf?

File: easybuild/easyblocks/t/tau.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+        # compose templates
+        mpi_tmpl = comp_opts + mpi_opts + opt_pkgs_opts + backend_opt
+        openmp_tmpl = comp_opts + openmp_opts + opt_pkgs_opts + backend_opt
+        hybrid_tmpl = comp_opts + openmp_opts + mpi_opts + opt_pkgs_opts + backend_opt
+
+        # number of iterations: # backends + 1 (for basic)
+        iter_cnt = len(self.cfg['extra_backends']) + 1
+
+        self.cfg['configopts'] = [mpi_tmpl, openmp_tmpl, hybrid_tmpl] * iter_cnt
```
Comment (wpoely86) [inline easybuild/easyblocks/t/tau.py]: maybe print (to debug) all of this?

File: easybuild/easyblocks/t/tau.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        backend = (['tau'] + self.cfg['extra_backends'])[self.variant_index // 3]
+        variant = ['mpi', 'openmp', 'hybrid'][self.variant_index % 3]
+        print_msg("starting with %s backend (%s variant)" % (backend, variant), log=self.log, silent=self.silent)
+
+    def prepare_step(self):
+        """Custom prepare step for Tau: check required dependencies and collect information on them."""
+        super(EB_TAU, self).prepare_step()
+
+        # install prefixes for selected backends
+        for dep in ['Scalasca', 'Score-P', 'Vampirtrace']:
```
Comment (wpoely86) [inline easybuild/easyblocks/t/tau.py]: put this together with the `KNOWN_BACKENDS` (as dict)

File: easybuild/easyblocks/t/tau.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            root = get_software_root(dep)
+            backend_name = dep.lower().replace('-', '')
```
Comment (wpoely86) [inline easybuild/easyblocks/t/tau.py]: or with this

File: easybuild/easyblocks/t/tau.py (hunk 1 above)
Comment (boegel) [inline easybuild/easyblocks/t/tau.py]: :)

File: easybuild/easyblocks/t/tau.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            if root:
+                self.opt_pkgs_opts += ' -%s=%s' % (opt, root)
+
+        # determine list of labels, based on selected (extra) backends, variants and optional packages
+        self.variant_labels = []
+        backend_labels = ['', '-epilog-scalasca-trace', '-scorep', '-vampirtrace-trace']
+        for backend, backend_label in zip(['tau'] + KNOWN_BACKENDS, backend_labels):
+            #if backend != 'tau' and backend not in self.cfg['extra_backends']:
+            if backend == 'tau' or backend in self.cfg['extra_backends']:
+
```
Comment (wpoely86) [inline easybuild/easyblocks/t/tau.py]: less whitespace in this part?

**Comment by geimer** [comment]
@boegel:  I noticed that you removed the `self.cfg['keeppreviousinstall'] = True` from the prepare step.  Was this intentional, i.e., is there another way to achieve the same behavior?  I'm asking because the PDT easyblock uses this technique as well...

**Comment by boegel** [comment]
@geimer I redefined `make_installdir` as `pass` instead, which is cleaner imho

File: easybuild/easyblocks/t/tau.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        avail_makefiles = ['Makefile.' + l for l in self.variant_labels]
+        if self.cfg['tau_makefile'] not in avail_makefiles:
+            raise EasyBuildError("Specified tau_makefile %s will not be available (only: %s)",
+                                 self.cfg['tau_makefile'], avail_makefiles)
+
+        # create install directory and make sure it does not get cleaned up again in the install step;
+        # the first configure iteration already puts things in place in the install directory,
+        # so that shouldn't get cleaned up afterwards...
+        self.log.info("Creating install dir %s before starting configure-build-install iterations", self.installdir)
+        mkdir(self.installdir, parents=True)
```
Comment (boegel) [inline easybuild/easyblocks/t/tau.py]: TODO: remove install dir if it's already there (e.g. because of forced reinstallation)

**Comment by boegel** [comment]
tested with https://github.com/hpcugent/easybuild-easyconfigs/pull/509, good to go now (finally...)

Thanks @geimer and @berndmohr!

**Pull request merged**